### PR TITLE
Optimize FLAC decoding

### PR DESCRIPTION
**Pull request by uklotzde**
Follow-up of #1660

### CODE REVIEW

File: src/sources/soundsourceflac.cpp [hunk 1]
```diff
@@ -345,25 +344,31 @@ FLAC__bool SoundSourceFLAC::flacEOF() {
...
+// Required at least for libFLAC 1.3.2. This workaround might become
+// obsolete once libFLAC is taking care of these issues internally.
+// https://bugs.launchpad.net/mixxx/+bug/1769717
+// https://hydrogenaud.io/index.php/topic,61792.msg559045.html#msg559045
+//
+// We will shift decoded samples left by (32 - m_bitsPerSample) to
+// get rid of the garbage in the most significant bits before scaling.
+// The range of decoded integer sample values then becomes
+// [-2 ^ 31, 2 ^ 31 - 1]. Afterwards this integer range needs to be
+// scaled to [-CSAMPLE_PEAK, CSAMPLE_PEAK).
```
Comment (daschuer) [inline src/sources/soundsourceflac.cpp]: [-2 ^ 31 - X, 2 ^ 31 - 1 - X] where X is the resolution  
scaled to [-CSAMPLE_PEAK, CSAMPLE_PEAK - X/2^31]
I would just remove the two last sentences, currently they are incorrect and later they will be confusing.

Comment (daschuer) [reply src/sources/soundsourceflac.cpp]: How about this: 

// We will shift decoded samples left by (32 - m_bitsPerSample) to
// get rid of the garbage in the most significant bits before scaling.
// The resulting full scale 32 bit signed sample is divided by 2 ^ 31 to 
// normalize it back to a floating point < CSAMPLE_PEAK.

**Review by daschuer — COMMENTED** [review]
Just two minor comments. LGTM 
Thank you for taking care.

**Comment by daschuer** [comment]
Thanks! LGTM

**Comment by daschuer** [comment]
waiting for CI

**Comment by ywwg** [comment]
Do we have automated tests for file decoding?  It would be cool to have a collection of "known-tricky" flac files that we could use to test the decoder while also making sure we don't break basic things like seeking.  (not needed in this PR)

**Comment by uklotzde** [comment]
For those tests we would need uncompressed reference signals for comparison and permissible deviations for each codec.

The current tests only compare the outputs of the decoder in various circumstances, e.g. with/without seeking before decoding. Our reference is the decoded signal itself and we won't notice if the decoding is entirely broken as long as it produces the same (distorted) signal.

**Comment by Pegasus-RPG** [comment]
I created a 1kHz 440Hz stereo reference track in src/test/soundFileFormats and a script to generate files in multiple formats for testing. Since it's a simple sine wave, testing for correct output can be done mathematically.

**Comment by ywwg** [comment]
> I created a 1kHz 440Hz stereo reference track in src/test/soundFileFormats and a script to generate files in multiple formats for testing. Since it's a simple sine wave, testing for correct output can be done mathematically.

I know in the case of the rubberband tests, we had to disable them because there was too much nondeterminism in the library.  (different results on win vs linux, for example).  FLAC may be ok because it's lossless, but compressed formats might differ slightly in the output.  I think that's what @uklotzde was getting at with respect to permissable deviations.

The current series of tests, testing the decoded stream, seem adequate.  A single sanity-check "this decoder is functioning properly" sounds good.  For flac and m4a it would be nice to have "decoding this set of bytes does not crash" for all these fixes we've been doing.

**Comment by daschuer** [comment]
I am not sure if it is worth the effort create a general sine wave test. The issue here, can be reproduced only with a very special reference file and all other hard nuts in other codecs are similar. 
So what we need is basically a encoded and decoded file for each of these cases. 
Maybe it is better to setup an independent "toxic" repository, which can be checked manually for a release. A good start is https://github.com/taglib/taglib/tree/master/tests/data by the way.

**Comment by Pegasus-RPG** [comment]
@asantoni Had a "library from hell" a few years ago that may be useful as well.

**Comment by Be-ing** [comment]
Ready for merge?

**Pull request merged**